### PR TITLE
Update cmake config to support new NCEP libraries (imported targets) on Tier-2 platforms

### PR DESCRIPTION
**Pull request by climbfuji**
## Description

This PR is a follow-up and clean-up for https://github.com/ufs-community/ufs-weather-model/pull/172.

- updates the modulefiles and cmake configurations for Tier-2 platforms in order to work with the new NCEP libraries (imported targets)
- cleans up cmake configurations (AVX2 flags, DEBUG and REPRO mode) and declares SIMDMULTIARCH flags as a real alternative to the AVX2 flags
- removes old build targets cheyenne.intel-mpi and supermuc-phase2.intel
- clean up of `build.sh`

Noting that for some platforms the responsibility for maintaining it may be handed off to somebody else and the directory layout/naming for the thirdparty and NCEP libraries may change as a result of it.

### Issue(s) addressed

n/a

## Testing

- compiles and runs on macosx.gnu, gaea.intel, jet.intel, stampede.intel, cheyenne.gnu, cheyenne.intel
- currently creating new baselines on cheyenne.gnu and cheyenne.intel using `rt.sh`
- need to run regression tests on tier-1 platforms and cheyenne.gnu and cheyenne.intel once these PRs are at the top of the commit queue

## Dependencies

Waiting on:
https://github.com/NOAA-EMC/fv3atm/pull/150
https://github.com/NOAA-EMC/NEMS/pull/74

### CODE REVIEW

**Comment by climbfuji** [comment]
Ready from my side.

**Comment by climbfuji** [comment]
Regression testing on cheyenne.intel and cheyenne.gnu: first create new baselines (Cheyenne did not yet have a 2020724 baseline), then verify against it: all tests pass, regression test logs updated in PR.

[rt_cheyenne_gnu_create.log](https://github.com/ufs-community/ufs-weather-model/files/4997613/rt_cheyenne_gnu_create.log)
[rt_cheyenne_gnu_verify.log](https://github.com/ufs-community/ufs-weather-model/files/4997614/rt_cheyenne_gnu_verify.log)
[rt_cheyenne_intel_create.log](https://github.com/ufs-community/ufs-weather-model/files/4997615/rt_cheyenne_intel_create.log)
[rt_cheyenne_intel_verify.log](https://github.com/ufs-community/ufs-weather-model/files/4997616/rt_cheyenne_intel_verify.log)

**Comment by climbfuji** [comment]
Regression testing on hera.intel and hera.gnu against existing baseline: all tests pass, logs updated in the PR.

[rt_hera_gnu.log](https://github.com/ufs-community/ufs-weather-model/files/4997714/rt_hera_gnu.log)
[rt_hera_intel.log](https://github.com/ufs-community/ufs-weather-model/files/4997716/rt_hera_intel.log)

**Comment by climbfuji** [comment]
Regression testing on orion.intel, wcoss_cray, wcoss_dell_p3 against existing baseline: all tests pass, logs updated in the PR.

**Review by DusanJovic-NOAA — APPROVED** [review]

File: cmake/GNU.cmake [hunk 1]
```diff
@@ -40,6 +39,9 @@ endif()
 if(AVX2)
     set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS}")
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
+elseif(SIMDMULTIARCH)
+    set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS}")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
```
Comment (aerorahul) [inline cmake/GNU.cmake:44]: What the the purpose of this?
A=A?

File: cmake/GNU.cmake [hunk 2]
```diff
@@ -1,24 +1,23 @@
 
 set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fcray-pointer -ffree-line-length-none -fno-range-check -fbacktrace")
 
+
 if(DEBUG)
     set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -Wall -O0 -ggdb -fno-unsafe-math-optimizations -frounding-math -fsignaling-nans -ffpe-trap=invalid,zero,overflow -fbounds-check")
```
Comment (aerorahul) [inline cmake/GNU.cmake:6]: This is not the correct way of adding `DEBUG` flags.
Please add DEBUG flags in CMAKE variable for e.g.
```
set(CMAKE_Fortran_FLAGS_DEBUG "-Wall -O0 -ggdb -fno-unsafe-math-optimizations -frounding-math -fsignaling-nans -ffpe-trap=invalid,zero,overflow -fbounds-check" )
```
Cmake will automatically add these flags when the build_type is DEBUG.
See 
https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html#variable:CMAKE_BUILD_TYPE

Please do this where `DEBUG` flags are set.

File: cmake/GNU.cmake [hunk 3]
```diff
@@ -1,24 +1,23 @@
...
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O2")
-endif()
-
-
-if(REPRO)
-    if(APPLE)
-        set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -O0 -ggdb")
-        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O0")
+    add_definitions(-DDEBUG)
+elseif(REPRO)
```
Comment (aerorahul) [inline cmake/GNU.cmake:9]: REPRO a `CMAKE_BUILD_TYPE`? If not, it should be defined as such.
Flags specific to `REPRO` should be defined as 
`CMAKE_<lang>_FLAGS_REPRO`

File: cmake/GNU.cmake [hunk 4]
```diff
@@ -40,6 +39,9 @@ endif()
 if(AVX2)
     set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS}")
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
```
Comment (aerorahul) [inline cmake/GNU.cmake:41]: What is the purpose of this? A=A.

File: cmake/Intel.cmake [hunk 5]
```diff
@@ -18,29 +18,33 @@ if(REPRO)
     add_definitions(-DREPRO)
     set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -qno-opt-dynamic-align")
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -qno-opt-dynamic-align")
+elseif(DEBUG)
+    add_definitions(-DDEBUG)
```
Comment (aerorahul) [inline cmake/Intel.cmake:22]: CMAKE_BUILD_TYPE is DEBUG.
Use appropriate `CMAKE_<lang>_FLAGS_DEBUG`

**Review by aerorahul — CHANGES_REQUESTED** [review]
There is incorrect use of defining compiler flags.
`CMAKE_<lang>_FLAGS_<BUILDTYPE>` should be defined for the appropriate `CMAKE_BUILD_TYPE`.
One should not "jam" those flags into the `CMAKE_<lang>_FLAGS` variable.
CMake knows what the BUILD TYPE is and will append the appropriate flags.

See: 
https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS.html
https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html#variable:CMAKE_BUILD_TYPE

File: cmake/GNU.cmake (hunk 1 above)
Comment (climbfuji) [reply cmake/GNU.cmake:44]: I didn't write this code.

File: cmake/GNU.cmake (hunk 4 above)
Comment (climbfuji) [reply cmake/GNU.cmake:41]: I didn't write this code.

File: cmake/Intel.cmake (hunk 5 above)
Comment (climbfuji) [reply cmake/Intel.cmake:22]: Please submit a follow-up PR to clean up the entire cmake build system. I am only making minimal changes to an existing cmake build system that already has DEBUG and REPRO defined. I don't think we want to make these changes now and rerun all the regression tests, but I may be mistaken.

File: cmake/GNU.cmake (hunk 1 above)
Comment (aerorahul) [reply cmake/GNU.cmake:44]: Understood.
Since there is cleaning e.g. `AVX2` etc. and adding flags, it is appropriate to comment.

Comment (aerorahul) [reply cmake/GNU.cmake:44]: This is an addition. GREEN.

File: cmake/Intel.cmake [hunk 6]
```diff
@@ -18,29 +18,33 @@ if(REPRO)
...
+        set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -qno-opt-dynamic-align")
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -qno-opt-dynamic-align")
     endif()
 endif()
 
 if(REPRO)
     set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -O2 -debug minimal -fp-model consistent -qoverride-limits -g -traceback")
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O2 -debug minimal")
+elseif(DEBUG)
+    set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -g -O0 -check -check noarg_temp_created -check nopointer -warn -warn noerrors -fp-stack-check -fstack-protector-all -fpe0 -debug -traceback -ftrapuv")
```
Comment (aerorahul) [inline cmake/Intel.cmake:42]: DEBUG flags are inappropriate in this CMAKE variable.

Comment (aerorahul) [reply cmake/Intel.cmake:42]: This is an addition. GREEN.

File: cmake/Intel.cmake (hunk 5 above)
Comment (aerorahul) [reply cmake/Intel.cmake:22]: This is new code added.
If you want to add something, adding it correctly won't impact older issues.
But, just building on older issues will make future clean-up harder.

**Comment by DusanJovic-NOAA** [comment]
@aerorahul This needs to be cleaned up. For example DEBUG flags was used (instead of CMAKE_BUILD_TYPE) because that  flag is used in GNU make build, so we wanted to maintain that during the transition and make maintenance of compile.sh scripts as similar as possible. There are many other redundant flags in both GNU and Intel configs (-qno-opt-dynamic-align for example). The original version of these files was just a rewrite of gnu make configuration files (ex. conf/configure.fv3.hera.intel). 

We keep postponing this cleanup until we fully transition to cmake (ie. remove GNU build entirely).

**Comment by climbfuji** [comment]
> @aerorahul This needs to be cleaned up. For example DEBUG flags was used (instead of CMAKE_BUILD_TYPE) because that flag is used in GNU make build, so we wanted to maintain that during the transition and make maintenance of compile.sh scripts as similar as possible. There are many other redundant flags in both GNU and Intel configs (-qno-opt-dynamic-align for example). The original version of these files was just a rewrite of gnu make configuration files (ex. conf/configure.fv3.hera.intel).
> 
> We keep postponing this cleanup until we fully transition to cmake (ie. remove GNU build entirely).

I'll have a quick chat with @aerorahul at 8.30 am MT to explain the origin for all this "build code". I'll invite you, too, in case you have time to attend.

File: cmake/Intel.cmake (hunk 5 above)
Comment (climbfuji) [reply cmake/Intel.cmake:22]: If I make the changes as you suggested, I'll break the existing builds. You'll see why in a minute when I walk you throuh the legacy code.

**Comment by climbfuji** [comment]
@aerorahul Could you please approve the PR based on our earlier discussion, noting that we will fix the issues you raised as part of our grand cleanup work? Thanks ...

**Review by aerorahul — APPROVED** [review]
Following enlightening offline conversation with @climbfuji and @DusanJovic-NOAA and recognizing the details and need for keeping the flags until a later time when the GNU build is decommissioned, I approve this PR.

I am leaving the comments here for reference.

**Review by junwang-noaa — APPROVED** [review]
I did not closely follow the cmake development, it looks to me the changes shows no impact on the results.  Maybe you can give some presentation on cmake build in ufs-weather-model once the cmake is able to replace the gnu build, thanks

**Comment by climbfuji** [comment]
Thanks, everyone!

**Comment by MinsukJi-NOAA** [comment]
@climbfuji, I see that in modulefiles/linux.gnu/fv3, library paths for NETCDF, ESMF, NCEPLIBS, and etc. have been removed.  Where should I specify them? -- I am using this fv3 file in my github actions workflow currently.

**Pull request merged**